### PR TITLE
[SDK] Add missing sal annotations in driverspecs.h and kernelspec.h

### PR DESCRIPTION
**Pull request by tkreuzer**
Fixes CORE-6706

### CODE REVIEW

File: sdk/include/psdk/driverspecs.h [hunk 1]
```diff
@@ -35,6 +34,12 @@
 #define _IRQL_uses_cancel_
 #define __drv_setsIRQL(irql)
 
+#undef _IRQL_inout_
+#define _IRQL_inout_                               _IRQL_saves_ _IRQL_restores_
+#define _IRQL_limited_to_(level)                   ASSERT(KeGetCurrentIrql() <= level); _Analysis_assume_(KeGetCurrentIrql() <= level);
```
Comment (katahiromz) [inline sdk/include/psdk/driverspecs.h:39]:
```suggestion
#define _IRQL_limited_to_(level) do { \
    ASSERT(KeGetCurrentIrql() <= (level)); \
    _Analysis_assume_(KeGetCurrentIrql() <= (level)); \
} while (0)
```

File: sdk/include/psdk/driverspecs.h [hunk 2]
```diff
@@ -78,63 +83,63 @@
...
-#define __kernel_code
-#define __kernel_driver
+#define __drv_isObjectPointer                       _Points_to_data_
+#define __drv_KMDF                                  __drv_Mode_impl(KMDF)
+#define __drv_maxFunctionIRQL(irql)                 _IRQL_always_function_max_(irql)
+#define __drv_maxIRQL(irql)                         _IRQL_requires_max_(irql)
+#define __drv_minFunctionIRQL(irql)                 _IRQL_always_function_min_(irql)
+#define __drv_minIRQL(irql)                         _IRQL_requires_min_(irql)
+#define __drv_Mode_impl(x)                          _Landmark_(x) __inline void __GENSYM(__SAL_dummy_)(void){}
+#define __drv_mustHold(kind)                        _When_(!_Arg_comp_(#kind,"Memory"), _Requires_lock_held_(_Curr_)) _When_(_Arg_comp_(#kind,"Memory"), _Kernel_requires_resource_held_(kind))
```
Comment (katahiromz) [inline sdk/include/psdk/driverspecs.h:100]: Too long line.

File: sdk/include/psdk/driverspecs.h [hunk 3]
```diff
@@ -78,63 +83,63 @@
...
+#define __drv_strictType(typename,mode)             _Pre_ _SA_annotes2(SAL_strictType, typename, mode)
+#define __drv_strictTypeMatch(mode)                 _Pre_ _SA_annotes1(SAL_strictTypeMatch,mode)
+#define __drv_unit(p)                               typedef int ___drv_unit_##p __GENSYM(__prefast_flag_kernel_driver_mode);
+#define __drv_useCancelIRQL                         _IRQL_uses_cancel_
+#define __drv_valueIs(arglist)                      _Post_ _SA_annotes1(SAL_return,arglist)
+#define __drv_WDM                                   __drv_Mode_impl(WDM)
+#define __drv_when(cond, annotes)                   _When_(cond, annotes)
+#define __internal_kernel_driver                    _Analysis_mode_(_Analysis_internal_kernel_driver_) _Analysis_mode_(_Analysis_local_leak_checks_)
+#define __kernel_code                               _Analysis_mode_(_Analysis_code_type_kernel_code_) _Analysis_mode_(_Analysis_local_leak_checks_)
+#define __kernel_driver                             _Analysis_mode_(_Analysis_code_type_kernel_driver_) _Analysis_mode_(_Analysis_local_leak_checks_)
```
Comment (katahiromz) [inline sdk/include/psdk/driverspecs.h:142]: Too long lines.

File: sdk/include/psdk/driverspecs.h [hunk 4]
```diff
@@ -143,13 +148,13 @@
...
     _Analysis_mode_(_Analysis_operator_new_throw_)
-#define __user_code
-#define __user_driver
-#define ___drv_unit_internal_kernel_driver
-#define ___drv_unit_kernel_code
-#define ___drv_unit_kernel_driver
-#define ___drv_unit_user_code
-#define ___drv_unit_user_driver
+#define __user_code                                	_Analysis_mode_(_Analysis_code_type_user_code_) _Analysis_mode_(_Analysis_local_leak_checks_)
+#define __user_driver                               _Analysis_mode_(_Analysis_code_type_user_driver_) _Analysis_mode_(_Analysis_local_leak_checks_)
```
Comment (katahiromz) [inline sdk/include/psdk/driverspecs.h:152]: Too long lines.

**Review by HBelusca — APPROVED** [review]

File: sdk/include/psdk/driverspecs.h (hunk 1 above)
Comment (HBelusca) [reply sdk/include/psdk/driverspecs.h:39]: These are not code lines to put e.g. within functions, but are code analyzer annotations that prefix the function prototypes: as such you cannot use the do-while(0) trick.

File: sdk/include/psdk/driverspecs.h (hunk 2 above)
Comment (tkreuzer) [reply sdk/include/psdk/driverspecs.h:100]: I don't think breaking lines makes it anymore readable.

File: sdk/include/psdk/driverspecs.h (hunk 1 above)
Comment (tkreuzer) [reply sdk/include/psdk/driverspecs.h:39]: No, these go into a function. You obviously can't call KeGetCurrentIrql() from outside of a function. I prefer keeping it as it is though, as MS does the same and the do/while could confuse the analyzer.

Comment (HBelusca) [reply sdk/include/psdk/driverspecs.h:39]: Ah, thanks for the correction :+1:

**Pull request merged**